spec/core/ics-004-channel-and-packet-semantics/LOCALHOST.md

```diff
@@ -0,0 +1,72 @@
+# Localhost Support in 04-channel
+
+## Rationale
+
+IBC provide the ICS-26 interface for two mutually-untrusted applications to negotiate a communication channel and start sending messages to each other. This interface is now widely adopted by applications (both smart contracts and static modules) so that they can speak to applications on remote chains. However, there is no way to reuse this interface to communicate with applications on the same chain. So applications that wish to communicate with counterparties on remote and local state machines must for the moment have two separate implementations to handle both cases. This is a dev-UX issue that we can solve by providing localhost functionality (analogous to localhost for the internet protocol).
+
+One approach is to implement this feature at the client layer. However this leads to a number of inefficiencies. Firstly, it requires the creation of a connection (handshake messages mediated by relayers) to be sent from the state machine back to itself in order to connect. It then requires that state that exists at the channel layer be routed all the way into the client so that it can then be inspected for verification purposes. Lastly, it does not immediately lend itself to future improvements that might make localhost packet flows atomic.
+
+The state that applications wish to verify about their counterparties are all stored in the channel store (i.e. `channelEnd`, `packet_commitment`, `receipt`, `acknowledgement`). Thus, the simplest approach is to add the special localhost handling here at the channel layer, rather than propogating it all the way down the stack. This provides the uniform interface that applications are looking for with minimal changes to the core handlers.
+
+## Technical Specification
+
+Implementations will reserve a special connectionID: `connection-localhost` which applications can build their channels on in order to communicate with a counterparty on the same state machine.
```

Comment:
Why not just `localhost`?

Reply:
In ibc-go there is connection id validation that requires the id to be in the format `connection-{N}`

Reply:
Interesting but there's no check on if the suffix is a number? Also would this ever get called on the localhost connection anyway?

Reply:
Ahh actually i was wrong, there is connection id validation that enforces the previous format i mentioned but that does not get called in the channel layer. The validation that gets called in `channel.ValidateBasic` seems to just enforce a character limit between 10-64 so `localhost` fails due to being only 9 chars.

https://github.com/cosmos/ibc-go/blob/3e8935e3334cac640a3f03b7a501a72aa6013624/modules/core/04-channel/types/channel.go#L68

Reply:
My preference would be `localhost` (adjust the connection validation to allow a lower limit), but I won't push very hard for it

Reply:
I kinda like it as-is